### ISSUE TITLE
A value typed graph

### ISSUE THREAD

**Issue by pepibumur** [tuist/tuist#1255]
## Graph 🕵️‍♀️
`Graph` is one of the core types of the codebase. It represents the projects' dependency graph which includes the nodes (i.e. targets, precompiled binaries, and packages), projects, and workspaces. Since its inception, we have evolved it to accommodate all the new features that we have been adding to Tuist. 
It has worked well, and has allowed us to build many great features that otherwise wouldn't have been possible; for example, making linking build phases and settings and implementation detail, or the `tuist graph` command.

As you might know, we have been talking a lot about **mappers lately and minimizing IO** to model the generation of projects as a combination of functions that end up passing down a descriptor to some executors that end up doing the IO and the writes of the Xcode projects and workspaces. The overview of that architecture is represented below:

![Untitled Diagram-Page-2 (1)](https://user-images.githubusercontent.com/663605/80125454-f5dee880-8591-11ea-8dda-91750bb22d75.png)

Note that we are aiming to do IO only when loading the graph, and writing the final project. Any IO that the mappers might need to do is modeled as a side effect descriptor that is executed at the end of the pipeline. One of the objects that we also pass down to the mappers that that we end up generating is the **graph**, **which is a reference type**. Since it's a reference type, nothing prevents the mappers or the linters to mutate it introducing unexpected side effects which might result in unexpected and non-deterministic results for the users. That's an unwanted state for Tuist to be in. Moreover, as we are adding more concurrency to the project, using reference types might result in threading race conditions that could cause Tuist to crash unexpectedly.

For those reasons, and before we continue going down the path of moving things to mappers, I'm proposing **making Graph a value type**

## Proposal

This is a proposal that I @kwridan and I came up with after doing a bit of pseudocode with the goal of coming up with value-typed graph that is efficient providing all of of the operations that we are currently doing with the current graph, as well as supporting future use cases like being able to replace nodes in the graph for the cache feature.

We'd define the following value types:

```swift
struct Target {
 /* Metadata */
}

enum Dependency {
    case xcframework(/* metadata */)
    case framework(/* metadata */)
    case library(/* metadata */)
    case cocoapods(/* paths */)
    case package(/* metadata */)
    case target(name: String, path: Path)
}

struct ValueGraph { 
    let projects: [AbsolutePath: Project]
    let targets: [AbsolutePath: [String: TargetDependency]]
    let dependencies: [Dependency: [Dependency]]
}
```
 
- **Target:** This will be very similar to the current `Target` in `TuistCore` with the differene that it'll be a value type, and it'll no longer have a `.dependencies` attribute.
- **Dependency:** It represents a node in the graph. Notice that the only case that doesn't contain metadata is the target. The reason being is that it'll allow us to change metadata from a target without having to traverse the whole graph to find all the references to the same target.
- **Graph:** This will be the new graph. In `projects` and `targets`, the key of the dictionary points to the directory where the project is defined. 

## How to get there?

Before anything else, we'd like to get feedback from all of you. Once we agree on the direction, here's a proposal on how to accomplish this migration:

- [x] **Introduce `ValueGraph` in `TuistCore`:** We'll just add the definition of the models with names that don't conflict with the current models. https://github.com/tuist/tuist/pull/1336
- [ ] **Via tests, implement the different mappers as a proof of concept:** This will serve as a validation that the graph that we are proposing will support ongoing and future efforts through mappers.  
- [ ] I**Introduce `GraphTraverser`** That will be responsible for providing convenient methods to get information from the graph. It'll have internal caching to avoid several passes through the edges. This might impact positively the performance of Tuist. This utility is only intended to be used by the generator so it'll be defined there.
- [ ] **Implement the logic that loads the manifests into a `ValueGraph` instance**
- [ ] **Replace usages of Graph with GraphTraverser:** If we implement the interface of `GraphTraverser` to be the same as the current `Graph` the process shouldn't be that painful. 

if all of that goes well, @kwridan suggested to go further on the loading side, and load manifests concurrently 🤯, which I think it's a brilliant idea. This refactor will make that possible.

cc @tuist/core

**Comment by fortmarek** [comment]
What a great proposal! This is promising and transforming `Graph` into a value type will make it much more easier to reason about and test plus, as you mention, the process can become even more concurrent.

I have just one question about `GraphTraverser` -> will `GraphTraverser` be a class or a value type? And if I understand it correctly, `GraphTraverser` should be just a wrapper around `ValueGraph`. What are the benefits of doing so? Is it because of the caching or is there more to it?

**Comment by kwridan** [comment]
Thanks for serializing that into an issue @pepibumur 👍 

> I have just one question about GraphTraverser -> will GraphTraverser be a class or a value type? And if I understand it correctly, GraphTraverser should be just a wrapper around ValueGraph. What are the benefits of doing so? Is it because of the caching or is there more to it?

Good question! Not sure of the final shape of `GraphTraverser` - wouldn't be surprised if it ends up internally looking similar to the current shape of `GraphNode`s with reference links to allow easy traversals - I guess the main motivation is that mapping itself doesn't take place on that structure which is optimized for querying rather than applying modifications. To help decide when its implemented, along side logical tests, we can add performance xctests to compare it to the existing graph look ups and help decide from there.

> Note that we are aiming to do IO only when loading the graph, and writing the final project. Any IO that the mappers might need to do is modeled as a side effect descriptor that is executed at the end of the pipeline. 

👍  - Worth a small update to the diagram to make it clear the generator itself doesn't do io, rather a separate phase afterwards that uses the descriptors to perform io + side effects.

in `ValueGraph` - should it be? 

```swift
struct TargetDependency {
  name: String
  path: Path
}

let targets: [AbsolutePath: [String: Target]]
let dependencies: [TargetDependency: [Dependency]]
```

I wonder if `cocoapods` is a true dependency vs a convenience to call `pod update` - (i.e. a side effect), may be an opportunity to remove it from the graph as it doesn't translate into any products like the other dependencies do and look for ways to model that differently. (e.g. recently in Tuist slack it was brought up - it being listed as a dependency made it seem like it could be defined in multiple projects which led to unexpected results / behavior 😞) 

It may help if we post some examples of `ValueGraph` might help demo it to see if its proposed structure works out or has any needed tweaks


```swift
let graph = ValueGraph(
   projects: [
       "/path/a": projectA,
       "/path/b": projectB,
   ],
  targets: [
      "/path/a", ["TargetA": targetA],
      "/path/a", ["TargetB": targetB],
      "/path/b", ["TargetC": targetC],
      "/path/b", ["TargetD": targetD],
  ]
  dependencies: [
     // ProjectA.TargetA depends on ProjectB.TargetC
      TargetDependency(name: "TargetA", path: "/path/a"): [
             .target(name: "TargetC", path: "/path/b"),
       ]
  ]
)
```

e.g. just doing this I realized we still need `targets` in `Project` otherwise we'd loose the order they are defined!

> if all of that goes well, @kwridan suggested to go further on the loading side, and load manifests concurrently 🤯, which I think it's a brilliant idea. This refactor will make that possible.

Not so much the manifests themselves 😛, the `Manifest > Model` conversion bit which can be slow due to globbing. I have a prototype of that here: [#1227 - `ProjectGenerator.load`](https://github.com/tuist/tuist/pull/1227/files#diff-b4a70cb82448e44c9f4f991d80f89f8aR137)

**Comment by ollieatkinson** [comment]
I am very interested in us pursuing a value based graph, it has many merits and opens up a lot of new possibilities in how we can use it. All my comments left here are just my thoughts - since we are still in early stages of discussion I hope I can help shape it.

```swift
enum Dependency {
    case xcframework(/* metadata */)
    case framework(/* metadata */)
    case library(/* metadata */)
    case cocoapods(/* paths */)
    case package(/* metadata */)
    case target(name: String, path: Path)
}
```

I'm not sure I think that an enum is a good representation here, it will make it very difficult to compose and might create quite a lot of boilerplate optional variables in order to get out common data.

It might be worth considering using protocols to define the types of metadata and allow concrete struct types to implement those.

e.g.

```swift
struct XCFrameworkDependency: ...
struct SwiftPackageDependency: ...
struct FrameworkDependency: ...
```

_where `...` represents the protocols they conform to_



> Via tests, implement the different mappers as a proof of concept: This will serve as a validation that the graph that we are proposing will support ongoing and future efforts through mappers.

Tests first is a very good idea, we have a lot of test cases covered by unit and fixtures. Adopting this new pattern is probably going to be tough, my suggestion here would be to implement an abstraction so we can switch between the two graphs and provide an experimental flag to Tuist's command line interface exposing it, so we can test it as well as the end-user and when we are happy we can remove the old code. A feature flag-esc approach if you like 😄 

> I wonder if cocoapods is a true dependency

I agree with @kwridan here that currently cocoa pods is _not_ a true dependency and it should not be represented as one - however this could change if we decide to support a deeper integration with cocoa pods.

From the proposed `ValueGraph` type it looks as if it's a cache, rather than a graph which means we start to loose the ability to traverse and perform operations on it as easy. I think we should keep a very generic view of the graph and use that to generate this representation. 

By doing so we keep the ability to calculate cycles very easily.
We can encode it and even hash incase we want to start using that to optimise loading.
By creating the encoding of the graph, we can provide that to users of Tuist so they can use it however they want, in a format they can understand.

```swift
struct Graph<V, E>: Collection, Codable where V: Hashable & Codable, E: Hashable & Codable {
    var vertices: [Vertex<V>]
    var edges: [[Edge<E>]]
}

extension Graph {
    public var startIndex: Int { 0 }
    public var endIndex: Int { vertices.count }
    public func index(after i: Int) -> Int { i + 1 }
    public subscript(i: Int) -> Vertex<V> { vertices[i] }
}

struct Vertex<T>: Codable where T: Hashable & Codable {
    let object: T
}

struct Edge<T>: Codable where T: Hashable & Codable {
    let from: Vertex<T>
    let to: Vertex<T>
}
```

> It may help if we post some examples of ValueGraph might help demo it to see if its proposed structure works out or has any needed tweaks

The proposed structure looks like a good start and easy to understand, but I do wonder if the whole structure could be flattened even further... something to think on.

Overall I'm really happy we are taking a look into this core part of Tuist, let's do something magical with it so we can get a _lot_ of functionality for very little cost 👍

**Comment by pepibumur** [comment]
> I wonder if cocoapods is a true dependency vs a convenience to call pod update - (i.e. a side effect), may be an opportunity to remove it from the graph as it doesn't translate into any products like the other dependencies do and look for ways to model that differently. (e.g. recently in Tuist slack it was brought up - it being listed as a dependency made it seem like it could be defined in multiple projects which led to unexpected results / behavior 😞)

I agree with this one. I'd in fact take the opportunity to remove it altogether from the interface and rather expose a hooks API that developers can use to run commands right after the project is generated. Having CocoaPods as a dependency has proven to confuse people because they can think they can integrate CocoaPods dependency graph into Tuist's dependency graph.

> e.g. just doing this I realized we still need targets in Project otherwise we'd lose the order they are defined!

@kwridan is the order important? If we sort them per type and then per name, does that make a difference? I'm leaning towards Tuist defining the order.

> I am very interested in us pursuing a value based graph, it has many merits and opens up a lot of new possibilities in how we can use it. All my comments left here are just my thoughts - since we are still in the early stages of discussion I hope I can help shape it.

That's a good point. Structs and protocols might be another good approach. I'm a bit concerned that it might lead to a lot of generics, which might introduce some other complications that we are not yet aware of. I'll give it a try.

> The ValueGraph type it looks like a cache rather than a graph which means we may start to lose the ability to traverse and perform operations.

What you presented there matches with the theory of what a graph could be like, but we have a slightly different case where we have a higher-order element, which are projects, that have references to nodes of the graph. The approach that @kwridan and I took to get to `ValueGraph` was starting from a practical standpoint *(i.e. what operations do we need to do efficiently in Tuist)* and think through different data structures that we could use to support all our use cases. In the end, it's a graph.

**Comment by kwridan** [comment]
> I agree with this one. I'd in fact take the opportunity to remove it altogether from the interface and rather expose a hooks API that developers can use to run commands right after the project is generated. Having CocoaPods as a dependency has proven to confuse people because they can think they can integrate CocoaPods dependency graph into Tuist's dependency graph.

👍 

> @kwridan is the order important? If we sort them per type and then per name, does that make a difference? I'm leaning towards Tuist defining the order.

Interesting one, I would image the order is somewhat needed, at least at the graph level, this could allow defining custom mappers to tweak the ordering if needed - worth gathering more feedback from others on this point. e.g. we tend to place the main or primary target of a project at the top with tests / mock / sample apps below it.

**Comment by kwridan** [comment]
## TL;DR

Now we have a `ValueGraph` and a few examples worth evaluating the usage before pressing ahead with replacing `Graph` throughout the codebase.

`ValueGraph` in its own right is absolutely fine, and having a separate `GraphTraverser`  is also a great idea. Taking a step back however to evaluate the motivation behind `ValueGraph` to address some of the short comings with mappers, there are still a few caveats/limitations we need to be aware off.

## Mappers

One of the big motivators for adopting a `ValueGraph` was mappers as such I will be focusing this post on this topic. 

As a quick reminder as to what this is, originally the pipeline for generation looked something like this

![Pipeline - Original](https://user-images.githubusercontent.com/11914919/88764510-fd9fc180-d16c-11ea-8fc1-e0a7bfed4313.png)

However there was a need to add additional behavior on top of generation, for example generating `infoPlist` files, modifying projects to include resource targets, and swap out projects all together for their cached `XCFramework` equivalent. Coupling this was the generator wasn't very feasible as such the idea of "mappers" was devised, that is operations that modify the input of the generator as well as produce side effects.

So the pipeline was modified to include mappers:

![load-manifests](https://user-images.githubusercontent.com/11914919/88764824-8585cb80-d16d-11ea-86ae-33991c16505a.png)

![model-mappers](https://user-images.githubusercontent.com/11914919/88765058-d4cbfc00-d16d-11ea-85c8-cbd5e6d77a29.png)

![graph-loader](https://user-images.githubusercontent.com/11914919/88765068-d85f8300-d16d-11ea-9c7f-95b6d4e06e43.png)

![graph-mapper](https://user-images.githubusercontent.com/11914919/88765079-dc8ba080-d16d-11ea-9601-bb6e3be1831f.png)

### Graph Mappers

The main area where we encountered issues with `Graph` was within mappers. It wasn't very "mappable" as it was optimized for look ups / traversal as opposed to allowing modifications.

For example the following attempts to introduce `GraphMappers` for items that modified, added or removed elements such as `dependencies`, `targets` or `projects` within the graph didn't quite work out. This was due to the models being cached in various locations within the `Graph`'s internal storage and as such most operations would yield some form of internal inconsistency.

examples:
- https://github.com/tuist/tuist/pull/1383
- https://github.com/tuist/tuist/pull/1357#discussion_r427048353
- https://github.com/tuist/tuist/pull/1206

Worth noting here, this isn't to say `Graph` is problematic, but rather mapping or modifying a `Graph` is.

### Model Mappers

To help mitigate the issues above, "Model Mappers"  (`ProjectMapping` and `TargetMapping`) were introduced in [#1378]( https://github.com/tuist/tuist/pull/1378).

Those are mappers that run on the models `Project` and `Target` ahead of making a graph. This enabled safely making modifications to model while ensuring the final graph was internaly consistent.

Examples:

- https://github.com/tuist/tuist/pull/1577
- https://github.com/tuist/tuist/pull/1567
- https://github.com/tuist/tuist/pull/1469
- https://github.com/tuist/tuist/pull/1467

One use case that is still not covered with model mappers is the ability to add / remove projects. For this we need to add something like `WorkspaceMapping` which takes in a collection of `[Projects]`. This is something that can be explored further if there is interest (e.g. attempting to implement a `CacheWorkspaceMapper`).

### Value Graph Mappers

`ValueGraph`  and `ValueGraphMapper` in theory should address the last missing use case as it does have access to all the projects as a collection.

However there are still some caveats to mapping `ValueGraph`

#### Duplicate information

- To add a target, for completeness the mapper would need to perform 2 operations
	- Update `graph.targets`
	- Update `targets`
- To add a new dependency for completeness the mapper would need to perform 3 operations
	- Update `graph.projects.targets`
	- Update `graph.targets`
	- Update `graph.dependencies`

#### Adding new external dependencies

- In a model mapper adding a new `XCFramework` dependency is one operation
   - Add `.xcframework(path: “/path/to/xcframework.xcframework”)` to the list of dependencies for that target - loading only happens once the graph is constructed at the very end
   - In a `ValueGraphMapper`
	- The 3 operations mentioned before are needed
	- Additionally,  `XCFrameworkLoader` is needed to load the xcframework metadata to add the appropriate `ValueGraph.dependency`

Loading additional data in a mapper is absolutely fine, it’s being highlighted here to compare and contrast.

### Mapper composability

One interesting exercise to think about is can mappers be composed together?

For example in model mappers, a `TargetMapper` can be implemented as a `ProjectMapper` - e.g. `TargetProjectMapper(mapper: targetMapper)`  - (see [`TargetProjectMapper`](https://github.com/tuist/tuist/blob/master/Sources/TuistCore/Graph/Mappers/ProjectMapper.swift#L25))

Can we make a `ProjectMapper` and `TargetMapper` implementation using `ValueGraphMapper` (e.g. `ProjectGraphMapper(mapper: projectMapper)` )?

From a cursory glance, the answer is possibly with a few gotchas :( 

- A `ProjectGraphMapper` would need to keep track of any newly added targets (and their dependencies) to update the internal representation as `targets` have their own storage in `ValueGraph`
- Similarly adding new dependencies within `ProjectGraphMapper` or `TargetGraphMapper` has two challenges
	- Updating the internal representation of dependencies
	- Loading new dependency metadata for external dependencies (e.g. `.framework` , `.library`, `.xcframework`)

**Comment by pepibumur** [comment]
Sorry that I haven't been able to answer earlier to your comment but I totally agree with you that need some further exploration before we migrate to value graph. I'm answering to some of your points below:

**Duplication of information in the value graph**
Once we move to the value graph, there'll be a single source of truth for dependencies, and that'll be `graph.dependencies`, `target.dependencies` and `project.targets.dependencies` will go away. Until we can get rid of `Graph`, it'll be 2 operations for adding a new target, and 3 operations for adding a dependency. However, once those attributes go away, they should only be one operation. In the case of adding a target it should be updating the `graph.targets` attribute that maps projects with targets, and in the case of adding dependencies we'll have to update `graph.dependencies`.

**Adding new external dependencies**
We can mimic what we have now, which as you pointed out, consists of adding the dependency and load any attributes that are necessary for the project generation. In the case of project generation for instance that's loading attributes like the architecture, or the type of linking. We can explore in the future whether we want this step to be further down the pipeline, for example right before generating the project. 

**One interesting exercise to think about is can mappers be composed together**
I like this idea, I'll prototype a bit with this to see if it's possible to achieve something like that.

**Comment by pepibumur** [comment]
Hey @kwridan,
I did a bit of prototyping [on this PR](https://github.com/tuist/tuist/pull/1656). I did not find any problems in initialising mappers using other mappers. The only thing though is that since we are making the graph the source of truth for dependencies and projects' targets, target and project mappers can't do things that they used to be able to do:

- Target mappers: modify the dependencies of a target.
- Project mappers: modify the targets of a project.

This means that some mappers might need to become a graph mappers or even better, what you have been suggesting for a while, have mappers that are applied before the graph is even formed.
